transportmiddleware/interface.go

```diff
@@ -0,0 +1,53 @@
+// Package transportmiddleware provides an interface for hooking into
+// outgoing yab requests.
+package transportmiddleware
+
+import (
+	"context"
+	"sync"
+
+	"github.com/yarpc/yab/transport"
+)
+
+var (
+	// stores the currently registered middleware
+	registeredMiddleware Interface
+
+	// serializes access to the currently registered middleware
+	registerLock sync.RWMutex
+)
+
+// Register sets the provided transport middleware to be used on future
+// calls to Apply(). Calls to Register() will overwrite previously registered
+// middlewares; that is, only one middleware is allowed at a time.
+// Returns a function to undo the change made by this call.
+func Register(newMW Interface) (restore func()) {
+	registerLock.Lock()
```

Comment:
(sorry for the back and forth about this topic)
i don't think we want a lock here. if we have multiple callers racing to call `Register`, the behaviour is completely undefined. that's bad behaviour, we'd prefer if that triggered a race.